mtags/src/main/scala/scala/meta/internal/pc/CompilerAccess.scala

```diff
@@ -214,8 +214,14 @@ abstract class CompilerAccess[Reporter, Compiler](
         { () =>
           if (!result.isDone()) {
             try {
+              val runsOnSameThread =
+                _compiler.presentationCompilerThread.isEmpty
               result.cancel(false)
               shutdownCurrentCompiler()
+              /* If running inside the executor, we need to reset the job queue
+               * Otherwise it will block indefinetely in case of infinite loops.
+               */
+              if (runsOnSameThread) jobs.reset()
```

Comment:
I changed the logic since, running interrupt work in an simple infinite loop I had, while more complex scenarios it broke. I will try and add a test later.